File: frontend/src/components/benchmark/content/johnson/implementations/johnson.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Johnson
{
    private class Edge
    {
        public int U { get; set; }
        public int V { get; set; }
        public int Weight { get; set; }
    }

    private List<Edge> edges = new List<Edge>();
    private Dictionary<int, HashSet<int>> adjacency = new Dictionary<int, HashSet<int>>();
    private HashSet<int> nodes = new HashSet<int>();

    public void AddEdge(int u, int v, int weight)
    {
        edges.Add(new Edge { U = u, V = v, Weight = weight });
        if (!adjacency.ContainsKey(u))
            adjacency[u] = new HashSet<int>();
        adjacency[u].Add(v);
        nodes.Add(u);
        nodes.Add(v);
    }

    public void BuildFromEdges(List<(int, int, int)> edgeList)
    {
        foreach (var (u, v, w) in edgeList)
        {
            AddEdge(u, v, w);
        }
    }

    private Dictionary<int, int> BellmanFord(int source)
    {
        var dist = new Dictionary<int, int>();
        foreach (int node in nodes)
            dist[node] = int.MaxValue;
        dist[source] = 0;

        for (int i = 0; i < nodes.Count - 1; i++)
        {
            bool updated = false;
            foreach (var edge in edges)
            {
                if (dist[edge.U] != int.MaxValue && dist[edge.U] + edge.Weight < dist[edge.V])
                {
                    dist[edge.V] = dist[edge.U] + edge.Weight;
                    updated = true;
                }
            }
            if (!updated) break;
        }

        foreach (var edge in edges)
        {
            if (dist[edge.U] != int.MaxValue && dist[edge.U] + edge.Weight < dist[edge.V])
            {
                return null;
            }
        }

        return dist;
    }

    private Dictionary<int, int> Dijkstra(int source, Dictionary<int, int> h)
    {
        var dist = new Dictionary<int, int>();
        foreach (int node in nodes)
            dist[node] = int.MaxValue;
        dist[source] = 0;

        var pq = new PriorityQueue<(int, int), int>();
        pq.Enqueue((0, source), 0);

        while (pq.Count > 0)
        {
            var (d, u) = pq.Dequeue();
            if (d != dist[u]) continue;

            if (!adjacency.ContainsKey(u)) continue;

            foreach (int v in adjacency[u])
            {
                int weight = 0;
                foreach (var edge in edges)
                {
                    if (edge.U == u && edge.V == v)
                    {
                        weight = edge.Weight;
                        break;
                    }
                }
                int newDist = dist[u] + weight + h[u] - h[v];
                if (newDist < dist[v])
                {
                    dist[v] = newDist;
                    pq.Enqueue((dist[v], v), dist[v]);
                }
            }
        }

        var result = new Dictionary<int, int>();
        foreach (int node in nodes)
        {
            result[node] = dist[node] - h[source] + h[node];
        }
        return result;
    }

    public Dictionary<int, Dictionary<int, int>> AllPairsShortestPaths()
    {
        int newNode = nodes.Max() + 1;
        foreach (int node in nodes)
        {
            AddEdge(newNode, node, 0);
        }

        var h = BellmanFord(newNode);
        if (h == null)
        {
            return null;
        }

        var result = new Dictionary<int, Dictionary<int, int>>();
        foreach (int node in nodes)
        {
            if (node == newNode) continue;
            result[node] = Dijkstra(node, h);
        }

        return result;
    }
}

File: frontend/src/components/benchmark/content/johnson/implementations/johnson.java
import java.util.*;

public class Johnson {
    private static class Edge {
        int u, v, weight;
        Edge(int u, int v, int weight) {
            this.u = u;
            this.v = v;
            this.weight = weight;
        }
    }

    private List<Edge> edges = new ArrayList<>();
    private Map<Integer, Set<Integer>> adjacency = new HashMap<>();
    private Set<Integer> nodes = new HashSet<>();

    public void addEdge(int u, int v, int weight) {
        edges.add(new Edge(u, v, weight));
        adjacency.putIfAbsent(u, new HashSet<>());
        adjacency.get(u).add(v);
        nodes.add(u);
        nodes.add(v);
    }

    public void buildFromEdges(List<int[]> edgeList) {
        for (int[] edge : edgeList) {
            addEdge(edge[0], edge[1], edge[2]);
        }
    }

    private Map<Integer, Integer> bellmanFord(int source) {
        Map<Integer, Integer> dist = new HashMap<>();
        for (int node : nodes) {
            dist.put(node, Integer.MAX_VALUE);
        }
        dist.put(source, 0);

        for (int i = 0; i < nodes.size() - 1; i++) {
            boolean updated = false;
            for (Edge edge : edges) {
                if (dist.get(edge.u) != Integer.MAX_VALUE &&
                    dist.get(edge.u) + edge.weight < dist.get(edge.v)) {
                    dist.put(edge.v, dist.get(edge.u) + edge.weight);
                    updated = true;
                }
            }
            if (!updated) break;
        }

        for (Edge edge : edges) {
            if (dist.get(edge.u) != Integer.MAX_VALUE &&
                dist.get(edge.u) + edge.weight < dist.get(edge.v)) {
                return null;
            }
        }

        return dist;
    }

    private Map<Integer, Integer> dijkstra(int source, Map<Integer, Integer> h) {
        Map<Integer, Integer> dist = new HashMap<>();
        for (int node : nodes) {
            dist.put(node, Integer.MAX_VALUE);
        }
        dist.put(source, 0);

        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[1]));
        pq.offer(new int[]{source, 0});

        while (!pq.isEmpty()) {
            int[] current = pq.poll();
            int u = current[0];
            int d = current[1];

            if (d != dist.get(u)) continue;

            if (!adjacency.containsKey(u)) continue;

            for (int v : adjacency.get(u)) {
                int weight = 0;
                for (Edge edge : edges) {
                    if (edge.u == u && edge.v == v) {
                        weight = edge.weight;
                        break;
                    }
                }
                int newDist = dist.get(u) + weight + h.get(u) - h.get(v);
                if (newDist < dist.get(v)) {
                    dist.put(v, newDist);
                    pq.offer(new int[]{v, dist.get(v)});
                }
            }
        }

        Map<Integer, Integer> result = new HashMap<>();
        for (int node : nodes) {
            result.put(node, dist.get(node) - h.get(source) + h.get(node));
        }
        return result;
    }

    public Map<Integer, Map<Integer, Integer>> allPairsShortestPaths() {
        int newNode = Collections.max(nodes) + 1;
        for (int node : nodes) {
            addEdge(newNode, node, 0);
        }

        Map<Integer, Integer> h = bellmanFord(newNode);
        if (h == null) {
            return null;
        }

        Map<Integer, Map<Integer, Integer>> result = new HashMap<>();
        for (int node : nodes) {
            if (node == newNode) continue;
            result.put(node, dijkstra(node, h));
        }

        return result;
    }
}